File: NotePal/Sources/NotePal/App/PetTheme.swift
import Foundation

enum PetTheme: String, CaseIterable, Identifiable {
    case newton
    case confucius
    case prayerIncenseBurner
    case specialA
    case specialB
    case tao

    var id: String {
        rawValue
    }

    var storedIDs: [String] {
        switch self {
        case .newton, .confucius, .prayerIncenseBurner:
            return [rawValue]
        case .specialA:
            return [rawValue, Self.joined("academic", "Special")]
        case .specialB:
            return [rawValue, Self.joined("wed", "ding", "Special")]
        case .tao:
            return [rawValue]
        }
    }

    var displayName: String {
        switch self {
        case .newton:
            return "Newton"
        case .confucius:
            return "孔子"
        case .prayerIncenseBurner:
            return "祈福香炉"
        case .specialA:
            return "Professor Ai"
        case .specialB:
            return "Dr.ZY & Dr.ZX"
        case .tao:
            return "Professor Tao"
        }
    }

    var resourceName: String {
        switch self {
        case .newton:
            return "Newton"
        case .confucius:
            return "Kongzi"
        case .prayerIncenseBurner:
            return "PrayerIncenseBurner"
        case .specialA:
            return "SpecialThemeA"
        case .specialB:
            return "SpecialThemeB"
        case .tao:
            return "SpecialThemeTao"
        }
    }

    var accessibilityLabel: String {
        switch self {
        case .newton:
            return "NotePal 默认形象"
        case .confucius:
            return "NotePal 经典形象"
        case .prayerIncenseBurner:
            return "NotePal 祈福香炉形象"
        case .specialA:
            return "NotePal 特殊形象 A"
        case .specialB:
            return "NotePal 特殊形象 B"
        case .tao:
            return "NotePal GIS 教授特殊形象"
        }
    }

    var isSpecial: Bool {
        switch self {
        case .newton, .confucius, .prayerIncenseBurner:
            return false
        case .specialA, .specialB, .tao:
            return true
        }
    }

    var publicResourceName: String? {
        isSpecial ? nil : resourceName
    }

    var encryptedResourceName: String? {
        isSpecial ? resourceName : nil
    }

    var sleepyPhrase: String {
        switch self {
        case .newton:
            return "我会安静思考一会儿。"
        case .confucius:
            return "吾少也贱，故多能鄙事。"
        case .prayerIncenseBurner:
            return "香烟静静升起。"
        case .specialA:
            return "我会安静待在这里。"
        case .specialB:
            return "我会安静待在这里。"
        case .tao:
            return "我先把图层关小声一点。"
        }
    }

    var completionPhrase: String {
        switch self {
        case .newton:
            return "完成一个待办，惯性被你打破了。"
        case .confucius:
            return "温故知新，又进一程。"
        case .prayerIncenseBurner:
            return "愿望又向前走了一步。"
        case .specialA:
            return "完成一个待办。"
        case .specialB:
            return "完成一个待办。"
        case .tao:
            return "又完成一个任务，空间索引都更清爽了。"
        }
    }

    var defaultGreeting: String {
        switch self {
        case .newton:
            return "从一颗苹果开始，也能想到整片天空。"
        case .confucius:
            return "学而时习之，不亦说乎。"
        case .prayerIncenseBurner:
            return "一缕清香，万事顺遂。"
        case .specialA:
            return "最近进展如何？"
        case .specialB:
            return "最近进展如何？"
        case .tao:
            return "今天的空间问题，先从尺度和位置讲起。"
        }
    }

    init?(storedValue: String) {
        switch storedValue {
        case Self.joined("academic", "Special"):
            self = .specialA
        case Self.joined("wed", "ding", "Special"):
            self = .specialB
        default:
            self.init(rawValue: storedValue)
        }
    }

    private static func joined(_ parts: String...) -> String {
        parts.joined()
    }
}

File: NotePal/Windows/NotePal.Windows/Models/PetTheme.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace NotePalWindows.Models;

public enum PetTheme
{
    Newton,
    Confucius,
    SpecialA,
    SpecialB
}

public static class PetThemeExtensions
{
    public static string ToRawValue(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "newton",
            PetTheme.Confucius => "confucius",
            PetTheme.SpecialA => "specialA",
            PetTheme.SpecialB => "specialB",
            _ => "newton"
        };
    }

    public static IReadOnlyList<string> StoredKeyNames(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.SpecialA => [theme.ToRawValue(), string.Concat("academic", "Special")],
            PetTheme.SpecialB => [theme.ToRawValue(), string.Concat("wed", "ding", "Special")],
            _ => [theme.ToRawValue()]
        };
    }

    public static string DisplayName(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "Default Theme",
            PetTheme.Confucius => "Classic Theme",
            PetTheme.SpecialA => "Special Theme A",
            PetTheme.SpecialB => "Special Theme B",
            _ => "NotePal"
        };
    }

    public static string ResourceName(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "Newton",
            PetTheme.Confucius => "Kongzi",
            PetTheme.SpecialA => "SpecialThemeA",
            PetTheme.SpecialB => "SpecialThemeB",
            _ => "Newton"
        };
    }

    public static bool IsSpecial(this PetTheme theme)
    {
        return theme is PetTheme.SpecialA or PetTheme.SpecialB;
    }

    public static string SleepyPhrase(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "我会安静思考一会儿。",
            PetTheme.Confucius => "吾少也贱，故多能鄙事。",
            PetTheme.SpecialA => "我会安静待在这里。",
            PetTheme.SpecialB => "我会安静待在这里。",
            _ => "我会安静待在这里。"
        };
    }

    public static string CompletionPhrase(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "完成一个待办，惯性被你打破了。",
            PetTheme.Confucius => "温故知新，又进一程。",
            PetTheme.SpecialA => "完成一个待办。",
            PetTheme.SpecialB => "完成一个待办。",
            _ => "完成一个待办。"
        };
    }

    public static string DefaultGreeting(this PetTheme theme)
    {
        return theme switch
        {
            PetTheme.Newton => "从一颗苹果开始，也能想到整片天空。",
            PetTheme.Confucius => "学而时习之，不亦说乎。",
            PetTheme.SpecialA => "最近进展如何？",
            PetTheme.SpecialB => "最近进展如何？",
            _ => "你好。"
        };
    }

    public static bool TryParseRawValue(string? rawValue, out PetTheme theme)
    {
        theme = PetTheme.Newton;
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return false;
        }

        switch (rawValue.Trim())
        {
            case "newton":
                theme = PetTheme.Newton;
                return true;
            case "confucius":
                theme = PetTheme.Confucius;
                return true;
            case "specialA":
                theme = PetTheme.SpecialA;
                return true;
            case "specialB":
                theme = PetTheme.SpecialB;
                return true;
            case var legacy when legacy == string.Concat("academic", "Special"):
                theme = PetTheme.SpecialA;
                return true;
            case var legacy when legacy == string.Concat("wed", "ding", "Special"):
                theme = PetTheme.SpecialB;
                return true;
            default:
                return Enum.TryParse(rawValue, ignoreCase: true, out theme);
        }
    }
}

public sealed class PetThemeJsonConverter : JsonConverter<PetTheme>
{
    public override PetTheme Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options)
    {
        if (reader.TokenType == JsonTokenType.String
            && PetThemeExtensions.TryParseRawValue(reader.GetString(), out var theme))
        {
            return theme;
        }

        if (reader.TokenType == JsonTokenType.Number && reader.TryGetInt32(out var value))
        {
            return Enum.IsDefined(typeof(PetTheme), value) ? (PetTheme)value : PetTheme.Newton;
        }

        return PetTheme.Newton;
    }

    public override void Write(Utf8JsonWriter writer, PetTheme value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToRawValue());
    }
}
